Delete entities and files

helenst opened the pull request:
Start by clicking 'delete' icon on the version of an entity - it'll lead through to the 'new version' page. From there, other files can be deleted/added.

Fixes #22 
Also should take care of #36

jonc125 commented:
I'm assuming this is ready to review @helenst ?

helenst commented:
@jonc125 Yes please :)

jonc125 commented on weblab/static/js/upload.js:
```diff
@@ -58,6 +56,19 @@ Upload.prototype = {
+      }
+      return false;
+    });
   },
```
Ah, that's OK because you don't show them the delete link for that file, just a 'cancel add' button effectively.

Interestingly, you can mark an existing file for deletion then upload a new file with the same name all within a single version change. Should we have a test that this does the right thing?

This is also possibly an unintended consequence of the fact that the check for a file upload having an existing name ignores files kept from previous versions - we should definitely fix that!

helenst replied:
I found a few bugs around marking for deletion and detecting existing uploaded files. It should now be possible to delete and replace a file in the same commit. (Might be useful if you have a local new copy of the file to upload - git will of course detect a deletion and re-addition as a change in the file)

jonc125 commented:
I added a couple of cosmetic changes to the branch as well.

helenst commented on weblab/entities/tests/test_views.py:
```diff
@@ -203,6 +235,73 @@ def test_create_model_version(self, user, client):
+        assert response.status_code == 302
+        assert response.url == '/entities/models/%d' % model.id
+        assert 'delete-file' in model.repo.tags
+        assert model.repo.head.commit.message == 'delete file1'
```
Agreed. It makes little sense to have a commit... I'll add a check.

helenst commented on weblab/entities/views.py:
```diff
@@ -195,6 +196,21 @@ def get_redirect_url(self, *args, **kwargs):
+    Delete an entity
+    """
+    model = Entity
+    raise_exception = True
```
It's part of `UserPassesTestMixin` which causes the view to raise a 403 error if the user doesn't pass the test rather than redirecting to login. Which seems to make sense, since deletion links aren't provided for those users anyway, it's not something that's likely to happen as part of the usual flow - a 'forbidden' response seems more appropriate (and they may already be logged in anyway). I'll add a comment.

helenst commented:
I think this has handled everything that was brought up, except for users acting simultaneously, which I've made into a separate issue for follow up work.

helenst commented:
i.e. ready for re-review @jonc125

jonc125 commented on weblab/entities/views.py:
```diff
@@ -239,33 +242,63 @@ def get_context_data(self, **kwargs):
-    def post(self, request, *args, **kwargs):
-        entity = self.get_object()
+    def add_and_delete_files(self, deletions, additions):
+        pass
```
Was this supposed to be implemented? 😄

helenst replied:
Oops... I had a refactor in mind but decided against it for now... Removed.

jonc125 commented on weblab/entities/views.py:
```diff
@@ -239,33 +242,63 @@ def get_context_data(self, **kwargs):
+                git_errors.append(e.stderr)
+
+        if git_errors:
+            # If there were any errors with adding or deleting files,
```
Will Django do something sensible with the temporary upload files in this case, keeping them for a resubmission attempt?

Also, what happens if the user abandons the new version? Does Django clean up automatically after an interval?

jonc125 replied:
@helenst can you answer the above queries?

helenst replied:
The temporary upload files will stay on the system (they're associated with the entity upon upload). There isn't a clean up yet, but probably should be. I'm not sure exactly at what point to decide it's appropriate to clean up an upload, since abandoning a version isn't a positive action by the user. Nightly job that cleans up anything older than 24 hours perhaps?

jonc125 replied:
That sounds sensible.

jonc125 commented on weblab/static/js/upload.js:
```diff
@@ -59,17 +56,30 @@ Upload.prototype = {
+    });
+  },
+
+  toggleDelete: function($row) {
```
Should this be `$tr`?

jonc125 commented on weblab/static/js/upload.js:
```diff
@@ -59,17 +56,30 @@ Upload.prototype = {
 
     $("form #entityversionfilestable").on('click', 'tr .action .delete-file', function() {
       var $tr = $(this).parents('tr');
+      return false;
```
This function doesn't appear to be doing anything?

jonc125 replied:
If I make it call `toggleDelete($tr)` (fixing the error below) then I can't delete multiple files in one go, because clicking the delete icon performs a GET request. (Why it should do this when the handler returns false I'm not sure!) If I leave it as-is, nothing happens at all when I click the delete icon.

helenst commented:
Argh - sorry. I did a little refactoring and thought it was all still working but it hadn't actually rebuilt the js. Sorry about that.

I've fixed this now. Turns out js errors were happening on click, so the handler couldn't get as far as returning false. I've fixed those. Can you see if it's all OK for you now?

jonc125 commented:
Well, deleting multiple files now works.

Alas, if I (accidentally in this case!) use an existing tag as the version field, I get a git command error traceback from the tag command. But by this point the commit has already happened. So I think we need a try-catch for this that moves HEAD back before the commit...

helenst commented:
@jonc125 That's a sensible idea. Will do...

jonc125 commented:
Works nicely. Now good to merge.
